src/lcm/dispatchers.py

```diff
@@ -40,42 +50,59 @@ def spacemap(func, dense_vars, sparse_vars, *, dense_first):
             above but there might be additional dimensions.
 
     """
+    # Check inputs
+    # ==================================================================================
+    if overlap := set(dense_vars).intersection(sparse_vars):
+        raise ValueError(
+            f"Dense and sparse variables must be disjoint. Overlap: {overlap}",
+        )
+
+    if duplicates := {v for v in dense_vars if dense_vars.count(v) > 1}:
+        raise ValueError(
+            f"Same argument provided more than once in dense variables: {duplicates}",
+        )
+
+    if duplicates := {v for v in sparse_vars if sparse_vars.count(v) > 1}:
+        raise ValueError(
+            f"Same argument provided more than once in sparse variables: {duplicates}",
+        )
+
+    # Preparations
+    # ==================================================================================
     func = allow_args(func)  # vmap cannot deal with keyword-only arguments
 
-    if not set(dense_vars).isdisjoint(sparse_vars):
-        raise ValueError("dense_vars and sparse_vars overlap.")
-
-    _all_variables = dense_vars + sparse_vars
-    if len(_all_variables) != len(set(_all_variables)):
-        raise ValueError("Same argument provided more than once.")
-
     signature = inspect.signature(func)
     parameters = list(signature.parameters)
 
-    positions = [parameters.index(cv) for cv in sparse_vars]
-    in_axes = []
-    for i in range(len(parameters)):
-        if i in positions:
-            in_axes.append(0)
+    positions_of_sparse_vars = [parameters.index(cv) for cv in sparse_vars]
+
+    in_axes_for_sparse_vars = []  # type: list[int | None]
+    for p in range(len(parameters)):
+        if p in positions_of_sparse_vars:
+            in_axes_for_sparse_vars.append(0)
         else:
-            in_axes.append(None)
+            in_axes_for_sparse_vars.append(None)
 
+    # Apply vmap for sparse and _product_map for dense variables
+    # ==================================================================================
     if not sparse_vars:
         vmapped = _product_map(func, dense_vars)
```

Comment:
Very confusing to have `productmap` and `_product_map` in the same module. Can we add something to the private function?

Reply:
`_base_productmap` with more info in the respective docstrings?

Reply:
Excellent, thanks!